### ISSUE TITLE
Make schema validator step and have prime-cli call it.

### ISSUE THREAD

**Issue by RickHawesUSDS** [CDCgov/prime-data-hub#236]
[ ] Check that each element has a Type

**Comment by MauriceReeves-usds** [comment]
@carlosfelix2 I was thinking that this might be another good thing you could work on. Currently we validate a schema by adding it to the `quick-test.sh` file and then running data through it, but it would be good to actually try to automate that and allow us to call `validate` as an option of the prime executable. 

### Schema Validation

I was thinking something like:

```
./prime validate --all-schemas
```

and that would load the metadata catalog as a first step (since schemas are loaded as part of the metadata) and then would attempt to run fake data through all the schemas in both the `CSV` and `HL7_BATCH` formats. The thing to keep in mind here is that some schemas don't have CSV fields defined, so it's possible to run data through the `CSV` format and have it return a zero-length document. That shouldn't necessarily be a failure.

I would also add an option to let us validate a single schema like:

```
./prime validate --schema <SCHEMA_NAME>
```

And in addition to say :

```
./prime validate --schema <SCHEMA_NAME> --output-format CSV
```

And have it provide warnings if an element in the schema doesn't have a `csvFields` value defined. This is only a warning because we include some fields into schemas so they can be consumed by mappers. When serializing elements we actually take two passes through the document to allow some fields to be loaded and THEN consumed by mappers on pass two. For example, you might want to do a 

```
- name: local_patient_name
   type: TEXT
   csvFields:
     - name: patient_name
   mapper: concat(patient_first_name, patient_last_name)

# standard fields consumed by mapper
- name: patient_first_name

- name: patient_last_name
```

Admittedly, this is a contrived example, but examples like this do exist. But it would be good to have it say something like:

```
WARNINGS:
- "patient_first_name" doesn't have a csvFields value defined. Is this on purpose?
- "patient_last_name" doesn't have a csvFields value defined. Is this on purpose?
```

Another area that we want this to help validate is the creation of `altValues` for a schema. We want to make sure that the codes an altValue maps to are actual codes defined for the value set.

### Organization and Receiver Validation
While the original ticket calls for just schema validation, I was also thinking it would be good to have this check organization and receiver files as well. The biggest areas where @jimduff-usds and I (mostly me) have tripped ourselves up are:

- Leaving the frequency for merging batches at 1440 per day for the prod file (where we would not want to batch every minute)
- Referencing jurisdictional filters that don't exist. I always write `or` when I mean `orEquals`.
- Not including a transport for a prod file (though this should be a warning, since it's okay to omit this since it will then just show up in the download site)

Being able to say:

```
./prime validate --organization-file <ORGANIZATION_FILE> --env prod 
```

It might also be good to be able to compare files and make sure that everything matches up for the same organization in two different files. This should only ever kick out warnings since we obviously expect differences in terms of merging and transport.

@RickHawesUSDS @jimduff-usds are there other kinds of validation it would be good to wrap into this?

**Comment by jimduff-usds** [comment]
Yes, I could go on all day with more QA ideas, but the above is more than enough to get a good start!

But...sorry....I just can't resist:   The one I'm thinking about is transforming data from our major senders (simple_report, strac and soon, waters) into all receivers.   I've done that manually as follows:

```
./prime data --input-fake 1000 --input-schema waters/waters-covid-19 --output ./junk/from-waters-to-all-states.csv
./prime data --input-fake 1000 --input-schema primedatainput/pdi-covid-19 --output ./junk/from-sr-to-all-states.csv
./prime data --input-fake 1000 --input-schema strac/strac-covid-19 --output ./junk/from-strac-to-all-states.csv

curl -X POST -H "client:strac"  -H "content-type:text/csv" --data-binary "@./junk/from-strac-to-all-states.csv" "http://localhost:7071/api/reports"
curl -X POST -H "client:waters"  -H "content-type:text/csv" --data-binary "@./junk/from-waters-to-all-states.csv" "http://localhost:7071/api/reports"
curl -X POST -H "client:simple_report"  -H "content-type:text/csv" --data-binary "@./junk/from-sr-to-all-states.csv" "http://localhost:7071/api/reports"

```

I'm about to put in a PR that makes the above actually work sans errors to all states and territories.

**Comment by MauriceReeves-usds** [comment]
I had another thought about a feature that should be added to this. The validator should give you the ability to compare orgs and receivers between environments, and optionally skip checking some sections.

### Some examples

*Compare All Organizations In a File Against Prod*
```
./prime validation compare --input organizations.yml --env prod --ignore MERGE,TRANSPORT
```

*Compare a Single Receiver in a File Against Prod*
```
./prime validation compare --input organizations.yml --env prod --ignore MERGE,TRANSPORT --names tx-doh.elr
```

*Compare a Single Receiver in a File Against Prod Ignoring Case*
```
./prime validation compare --input organizations.yml --env prod --ignore MERGE,TRANSPORT --names tx-doh.elr --ignore-case
```

The validation compare should load objects into memory and doing a simple comparison, listing any differences between the two sets of organization data. Also, any organizations/receivers in the local input that aren't in the `-env` provided should be listed as a difference, and vice versa.

In the ideal case, putting aside differences in `merge`, `transport`, and `meta`, what's in the `organizations-prod.yml` folder should  match 100%. This would help us catch any differences.
